ddtrace/tracer/post_processor.go

```diff
@@ -0,0 +1,113 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022 Datadog, Inc.
+
+package tracer
+
+import (
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+var _ ddtrace.ReadWriteSpan = (*readWriteSpan)(nil)
+
+// readWriteSpan wraps span and implements the ddtrace.ReadWriteSpan interface.
+type readWriteSpan struct {
+	*span
+}
+
+// Tag returns the tag value held by the given key.
+func (s readWriteSpan) Tag(key string) interface{} {
+	s.Lock()
+	defer s.Unlock()
+
+	switch key {
+	// String.
+	case ext.SpanName:
+		return s.Name
+	case ext.ServiceName:
+		return s.Service
+	case ext.ResourceName:
+		return s.Resource
+	case ext.SpanType:
+		return s.Type
+	// Bool.
+	case ext.AnalyticsEvent:
+		return s.Metrics[ext.EventSampleRate] == 1.0
+	case ext.ManualDrop:
+		return s.Metrics[keySamplingPriority] == -1
+	case ext.ManualKeep:
+		return s.Metrics[keySamplingPriority] == 2
+	// Metrics.
+	case ext.SamplingPriority, keySamplingPriority:
+		if val, ok := s.Metrics[keySamplingPriority]; ok {
+			return val
+		}
+		return nil
+	}
+	if val, ok := s.Meta[key]; ok {
+		return val
+	}
+	if val, ok := s.Metrics[key]; ok {
+		return val
+	}
+	return nil
+}
+
+// IsError reports wether s is an error.
+func (s readWriteSpan) IsError() bool {
+	s.Lock()
+	defer s.Unlock()
+
+	return s.Error == 1
+}
+
+// SetOperationName is not allowed in the processor and will not modify the operation name.
+func (s readWriteSpan) SetOperationName(operationName string) {
+	log.Debug("Modifying the operation name in the processor is not allowed")
+}
+
+// SetTag adds a set of key/value metadata to the span. Setting metric aggregator tags
+// (name, env, service, version, resource, http.status_code and keyMeasured) or modifying
+// the sampling priority in the processor is not allowed.
+func (s readWriteSpan) SetTag(key string, value interface{}) {
```

Comment:
I recommend that you think very well before _not_ using a pointer for the receiver.
```suggestion
func (s *readWriteSpan) SetTag(key string, value interface{}) {
```
In this scenario, the lock on the next line is copied, and you are locking a different lock than the `*span` you are working with, which can cause a race condition.

Reply:
Ah, missed this. Thanks for catching this, will make sure to think very well before not using a pointer for the receiver in the future! Changed in last commit.